sdks/java/io/hadoop-input-format/src/main/java/org/apache/beam/sdk/io/hadoop/inputformat/HadoopInputFormatIO.java

```diff
@@ -0,0 +1,967 @@
+import org.apache.beam.sdk.transforms.display.DisplayData;
+import org.apache.beam.sdk.util.CoderUtils;
+import org.apache.beam.sdk.values.KV;
+import org.apache.beam.sdk.values.PBegin;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.TypeDescriptor;
+import org.apache.hadoop.conf.Configurable;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.io.Writable;
+import org.apache.hadoop.mapreduce.InputFormat;
+import org.apache.hadoop.mapreduce.InputSplit;
+import org.apache.hadoop.mapreduce.Job;
+import org.apache.hadoop.mapreduce.RecordReader;
+import org.apache.hadoop.mapreduce.TaskAttemptID;
+import org.apache.hadoop.mapreduce.task.TaskAttemptContextImpl;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.auto.value.AutoValue;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Function;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Lists;
+
+/**
+ * A read transform for any data source which implements Hadoop InputFormat. For example- Cassandra,
+ * Elasticsearch, HBase, Redis, Postgres etc.
+ * <p>
+ * The Hadoop InputFormat source returns a set of {@link org.apache.beam.sdk.values.KV} key-value
+ * pairs returning a {@code PCollection<KV>}.
+ * <p>
+ * Hadoop Configuration is mandatory for reading data using Hadoop InputFormat source.
+ * <p>
+ * Hadoop {@link org.apache.hadoop.conf.Configuration Configuration} object will have to be set with
+ * following properties without fail:
+ * <ul>
+ * <li>mapreduce.job.inputformat.class : Signifies InputFormat class required to read data from
+ * source. For example, if user wants to read data from Cassandra using CqlInputFormat, then
+ * 'mapreduce.job.inputformat.class' property must be set to CqlInputFormat.class.</li>
+ * <li>key.class : Signifies output key class of InputFormat specified in
+ * 'mapreduce.job.inputformat.class'. For example, if class CqlInputFormat is specified in
+ * 'mapreduce.job.inputformat.class', then 'key.class' property must be set to Long.class.</li>
+ * <li>value.class : Signifies output value class of InputFormat specified in
+ * 'mapreduce.job.inputformat.class'. For example, if class CqlInputFormat is specified in
+ * 'mapreduce.job.inputformat.class', then 'key.class' property must be set to Row.class.</li>
+ * </ul>
+ * For example:
+ * 
+ * <pre>
+ * <tt>
+ *   Configuration myHadoopConfiguration = new Configuration(false);
+ *   // Set Hadoop InputFormat, key and value class in configuration
+ *   myHadoopConfiguration.setClass("mapreduce.job.inputformat.class", InputFormatClass, InputFormat.class);
+ *   myHadoopConfiguration.setClass("key.class", InputFormatKeyClass, Object.class);
+ *   myHadoopConfiguration.setClass("value.class", InputFormatValueClass, Object.class);
+ * </tt>
+ * </pre>
+ * <p>
+ * Key/value translation is a mechanism provided to user for translating InputFormat's key/value
```

Comment:
You might also add a comment that this is optional and that you don't need to specify both key & value.